specs/agents/tracing-spans-compress.md

```diff
@@ -0,0 +1,211 @@
+- Downstream effects like reducing impact on APM Server, ES storage, and UI performance
+
+
+### Consecutive-Exact-Match compression strategy
+
+One of the biggest sources of excessive data collection are n+1 type queries and repetitive requests to a cache server.
+This strategy detects consecutive spans that hold the same information (except for the duration)
+and creates a single [composite span](#composite-span).
+
+```
+[                              ]
+GET /users
+ [] [] [] [] [] [] [] [] [] []
+ 10x SELECT FROM users
+```
+
+Two spans are considered to be an exact match if they are of the [same kind](#consecutive-same-kind-compression-strategy) and if their span names are equal:
+- `type`
+- `subtype`
+- `destination.service.resource`
+- `name`
+
+### Consecutive-Same-Kind compression strategy
+
+Another pattern that often occurs is a high amount of alternating queries to the same backend.
+Especially if the individual spans are quite fast, recording every single query is likely to not be worth the overhead.
+
+```
+[                              ]
+GET /users
+ [] [] [] [] [] [] [] [] [] []
+ 10x Calls to mysql
+```
+
+Two spans are considered to be of the same type if the following properties are equal:
+- `type`
+- `subtype`
+- `destination.service.resource`
+
+```java
+boolean isSameKind(Span other) {
+    return type == other.type
+        && subtype == other.subtype
+        && destination.service.resource == other.destination.service.resource
+}
+```
+
+When applying this compression strategy, the `span.name` is set to `Calls to $span.destination.service.resource`.
+The rest of the context, such as the `db.statement` will be determined by the first compressed span, which is turned into a composite span.
+
+
+#### Configuration option `same_kind_compression_max_duration`
+
+Consecutive spans to the same destination that are under this threshold will be compressed into a single composite span.
+This option does not apply to [composite spans](#composite-span).
+This reduces the collection, processing, and storage overhead, and removes clutter from the UI.
+The tradeoff is that the statement information will not be collected. 
+
+|                |          |
+|----------------|----------|
+| Type           | `duration`|
+| Default        | `5ms`    |
+| Dynamic        | `true`   |
+
+### Composite span
+
+Compressed spans don't have a physical span document.
+Instead, multiple compressed spans are represented by a composite span.
+
+#### Data model
+
+The `timestamp` and `duration` have slightly similar semantics,
+and they define properties under the `composite` context.
+
+- `timestamp`: The start timestamp of the first span.
+- `duration`: The sum of durations of all spans.
+- `composite`
+    - `count`: The number of compressed spans this composite span represents.
+      The minimum count is 2 as a composite span represents at least two spans.
+    - `end`: The end timestamp of the last compressed span.
```

Comment:
Could we maybe use the [ECS `event.end` field](https://www.elastic.co/guide/en/ecs/current/ecs-event.html#field-event-end) for this?

Reply:
What would be the benefit of using that field? I'm just concerned that we would spread the information that belongs together (all the subfields of composite) across different logical groupings, so the semantics might become difficult to understand.

Reply:
This question came from https://github.com/elastic/apm-server/pull/5661#discussion_r667672013. Generally I'd like to move more of our schema over to ECS, to standardise what we can.

With the proposal as it stands, we would have:
 - `@timestamp` (ECS)
 - `span.duration.us` (non-ECS)
 - `span.composite.end` (non-ECS)

I'd like to replace both `span.duration.us` and `transaction.duration.us` with the ECS field [`event.duration`](https://www.elastic.co/guide/en/ecs/current/ecs-event.html#field-event-duration). If we did that, `span.composite.end` would be the odd field out, when it comes to timing-related properties.

Having said all that, I think (as @estolfo [pointed out](https://github.com/elastic/apm-server/pull/5661#discussion_r667949147)) that the semantics don't quite match, since `start+duration` is not necessarily equal to `end`.

Reply:
Just to clarify, you propose to have the following structure, right?:
```
{
   "name": "SELECT FROM product_types WHERE user_id=?",
   "timestamp": 1613575086000000,
   "event": {
        "duration": 3781,
        "end": 1613575093183900
   }
   "composite": {
       	"count": 10,
        "exact_match": true
    }
}
```
So, we would keep the `count` field within composite (as there is no count on event IIUC).
And the semantics then are:

- `event.end` - `timestamp` : Overall duration of the composite span (including the gaps between individual compressed spans).
- `event.duration` : Net duration of the compressed spans (sum of the durations of compressed spans)
- `event.duration` / `composite.count` : Average duration of the compressed spans.

I'n fine with that!  👍

Reply:
@AlexanderWert that's what I had in mind, yes. Only question for me now is whether we're breaking ECS semantics, which I think maybe we are. I'm seeking clarification from the ECS team.

Reply:
Just to make sure we are on the same page - is [the last suggestion](https://github.com/elastic/apm/pull/453#discussion_r668453491) for APM Server Intake API or for storage (i.e., Elasticsearch schema)?

WDYT about the [change to APM Server Intake API proposed above](https://github.com/elastic/apm/pull/453#discussion_r664324362):

> I wonder if it's worth going with the same model we have in breakdown metrics
> 
> * `sum.us` (a new property in `composite` object) to hold the sum of durations of all compressed spans
> * `duration` holding gross duration (i.e., <last compressed span's end timestamp> - <first compressed span's start timestamp>)
> * `end` being unnecessary anymore (effectively being replace by `sum.us`)
> 
> The pros of the new model are:
> 
> * `sum.us` and `count` is what we already have in breakdown metrics so it's more consistent
> * Compressed-spans unaware UI will display a composite span using its gross duration which is closer to what we want although that might be debatable :)

This way there's no `end` field in Intake API and if APM Server calculates it from `timestamp` and `duration` and store (although I'm not sure why) that will also work.

Reply:
ECS docs states the following:

> `event.duration` : Duration of the event in nanoseconds. If event.start and event.end are known this value **should be the difference between the end and start time**.

--> so seems like this would break this assumption

Alternative could be to have an explicit field for the net duration for the composite span (which would have a clear semantical difference as opposed to `event.duration` ):

```
{
   "name": "SELECT FROM product_types WHERE user_id=?",
   "timestamp": 1613575086000000,
   "event": {
        "duration": 3781,
   }
   "composite": {
        "net_duration": 3452
       	"count": 10,
        "exact_match": true
    }
}
```

Reply:
⬆️ actually the same as Sergey proposed, just with a different name

Reply:
@SergeyKleyman

> Just to make sure we are on the same page - is the last suggestion for APM Server Intake API or for storage (i.e., Elasticsearch schema)?

Elasticsearch schema, but I guess we should keep it somewhat consistent.

> WDYT about the change to APM Server Intake API proposed above:

I like that -- both for the intake API and for the Elasticsearch schema.

Reply:
> I like that -- both for the intake API and for the Elasticsearch schema.

[Felix was not against](https://github.com/elastic/apm/pull/453#discussion_r665128923) this change :) so I will apply it.

Reply:
Done